engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java

```diff
@@ -0,0 +1,57 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.interop.*;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.dsl.Builtin;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+/**
+ * A wrapper that allows to turn an Enso callback providing elements into a polyglot Array.
+ *
+ * <p>This allows creation of arrays (and with them, vectors) using non-standard storage - for
+ * example exposing rows of a Table without copying any data.
+ */
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+@Builtin(pkg = "immutable", stdlibName = "Standard.Base.Data.Array_Proxy.Array_Proxy")
+public class ArrayProxy implements TruffleObject {
+  private final long length;
+  private final Object at;
+
+  @Builtin.Method(description = "Creates an array backed by a proxy object.")
+  public ArrayProxy(long length, Object at) {
```

Comment:
Shouldn't `at` be a `IntFunction<Value>` or similar?

Reply:
No, it cannot be `IntFunction<Value>`. Such code would be appropriate in _host interop Java_, however here we are writing  a builtin and it needs to be written according to Truffle API - e.g. using `InteropLibrary` rather than `Value`.

Reply:
ah ok - thanks for explaining.

Reply:
Although I think @jdunkerley you have raised a fair point that it is good to check in the constructor that the `Object at` is really an executable function.

I'll add
```
if (CompilerDirectives.inInterpreter()) {
      if (!InteropLibrary.getUncached().isExecutable(at)) {
        throw new IllegalStateException("Array_Proxy needs an executable callback, but got: " + at);
      }
    }
```

inspired by how Vector checks its `storage`.